vgen/src/main/java/io/gapi/vgen/client/ClientConfigGapicLanguageProvider.java

```diff
@@ -0,0 +1,100 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gapi.vgen.client;
+
+import com.google.api.tools.framework.model.Interface;
+import com.google.api.tools.framework.model.Method;
+import com.google.api.tools.framework.model.Model;
+import com.google.api.tools.framework.model.ProtoFile;
+import com.google.api.tools.framework.snippet.Doc;
+import com.google.api.tools.framework.snippet.SnippetSet;
+import com.google.api.tools.framework.tools.ToolUtil;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
+
+import io.gapi.vgen.ApiConfig;
+import io.gapi.vgen.GapicContext;
+import io.gapi.vgen.GapicLanguageProvider;
+import io.gapi.vgen.GeneratedResult;
+import io.gapi.vgen.SnippetDescriptor;
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class ClientConfigGapicLanguageProvider implements GapicLanguageProvider {
+  
+  private Model model;
+  private ApiConfig apiConfig;
+  
+  public ClientConfigGapicLanguageProvider(Model model, ApiConfig apiConfig) {
+    this.model = model;
+    this.apiConfig = apiConfig;
+  }
+
+  @Override
+  public Model getModel() {
+    return model;
+  }
+
+  @Override
+  public <Element> void output(
+      String outputPath, Multimap<Element, GeneratedResult> elements)
+      throws IOException {
+    Map<String, Doc> files = new LinkedHashMap<>();
+    for (GeneratedResult generatedResult : elements.values()) {
+      files.put(generatedResult.getFilename(), generatedResult.getDoc());
+    }
+    
+    ToolUtil.writeFiles(files, outputPath);
+  }
+
+  /**
+   * The path to the root of snippet resources.
+   */
+  private static final String SNIPPET_RESOURCE_ROOT =
+      ClientConfigGapicLanguageProvider.class.getPackage().getName().replace('.', '/');
+  
+  @SuppressWarnings("unchecked")
+  @Override
+  public GeneratedResult generateCode(Interface service, SnippetDescriptor snippetDescriptor) {
+    ClientConfigSnippetSet<Interface> snippets =
+        SnippetSet.createSnippetInterface(
+            ClientConfigSnippetSet.class,
+            SNIPPET_RESOURCE_ROOT,
+            snippetDescriptor.getSnippetInputName(),
+            ImmutableMap.<String, Object>of(
+                "context", new GapicContext(model, apiConfig)));
+
+    String outputFilename = snippets.generateFilename(service).prettyPrint();
+
+    // Generate the body, which will collect the imports.
+    Doc body = snippets.generateBody(service);
+
+    return GeneratedResult.create(body, outputFilename);
+  }
+
+  // There are no fragments for the client config
+  @Override
+  public GeneratedResult generateFragments(Method method, SnippetDescriptor snippetDescriptor) {
+    return null;
```

Comment:
Other languages don't provide generateDoc; I see subclasses of GapicLanguageProvider as implementing the methods that are pertinent to that particular GAPIC. Also this requires changing the CodeGenerator, which expects a GapicLangaugeProvider. We _could_ refactor so that the CodeGenerator instead takes a base class of GapicLangaugeProvider, but I'm not sure it's worth the effort because it doesn't make things particularly simpler or more readable in this case. What do you think?